tmol/pack/rotamer/build_rotamers.py

```diff
@@ -0,0 +1,1043 @@
+
+    for i, samples in enumerate(chi_samples):
+        rot_counter_for_rt = (
+            torch.arange(samples[1].shape[0], dtype=torch.int64, device=device)
+            - rt_nrot_offsets[i][samples[1].to(torch.int64)]
+        )
+        numpy.testing.assert_array_less(
+            rot_counter_for_rt.cpu().numpy(), max_n_rotamers_per_rt
+        )
+
+    sort_rt_for_rotamer = torch.cat(
+        [
+            samples[1].to(torch.int64) * len(chi_samples) * max_n_rotamers_per_rt
+            + i * max_n_rotamers_per_rt
+            + torch.arange(samples[1].shape[0], dtype=torch.int64, device=device)
+            - rt_nrot_offsets[i][samples[1].to(torch.int64)]
+            for i, samples in enumerate(chi_samples)
+        ]
+    )
+    sampler_for_rotamer_unsorted = torch.cat(
+        [
+            torch.full((samples[1].shape[0],), i, dtype=torch.int64)
+            for i, samples in enumerate(chi_samples)
+        ]
+    )
+    sort_ind_for_rotamer = torch.argsort(sort_rt_for_rotamer)
+    sort_rt_for_rotamer_sorted = sort_rt_for_rotamer[sort_ind_for_rotamer]
+    uniq_sort_rt_for_rotamer = torch.unique(sort_rt_for_rotamer_sorted)
+
+    assert uniq_sort_rt_for_rotamer.shape[0] == sort_rt_for_rotamer_sorted.shape[0]
+
+    sampler_for_rotamer = sampler_for_rotamer_unsorted[sort_ind_for_rotamer]
+
+    all_rt_for_rotamer = torch.cat([samples[1] for samples in chi_samples])[
+        sort_ind_for_rotamer
+    ]
+
+    max_n_chi_atoms = max(samples[2].shape[1] for samples in chi_samples)
+    all_chi_atoms = torch.full(
+        (n_rotamers, max_n_chi_atoms), -1, dtype=torch.int32, device=device
+    )
+    all_chi = torch.full(
+        (n_rotamers, max_n_chi_atoms), -1, dtype=torch.float32, device=device
+    )
+    offset = 0
+    for samples in chi_samples:
+        assert samples[2].shape[0] == samples[3].shape[0]
+        all_chi_atoms[
+            offset : (offset + samples[2].shape[0]), : samples[2].shape[1]
+        ] = samples[2]
+        all_chi[
+            offset : (offset + samples[2].shape[0]), : samples[3].shape[1]
+        ] = samples[3]
+        offset += samples[2].shape[0]
+
+    all_chi_atoms = all_chi_atoms[sort_ind_for_rotamer]
+    all_chi = all_chi[sort_ind_for_rotamer]
+
+    # ok, now we need to figure out how many rotamers each rt is getting.
+    n_rots_for_rt = toolz.reduce(torch.add, [samples[0] for samples in chi_samples])
+
+    return (
+        n_rots_for_rt,
+        sampler_for_rotamer,
+        all_rt_for_rotamer,
+        all_chi_atoms,
+        all_chi,
+    )
+
+
+@validate_args
+def create_dof_inds_to_copy_from_orig_to_rotamers(
+    poses: PoseStack,
+    task: PackerTask,
+    samplers,  # : Tuple[ChiSampler, ...],
+    rt_for_rot: Tensor[torch.int64][:],
+    block_type_ind_for_rot: Tensor[torch.int64][:],
+    sampler_for_rotamer: Tensor[torch.int64][:],
+    n_dof_atoms_offset_for_rot: Tensor[torch.int32][:],
+) -> Tuple[Tensor[torch.int64][:], Tensor[torch.int64][:]]:
```

Comment:
wonder if it makes sense to move all the indexing fns to an indexing class?

Reply:
What would that look like?

Reply:
(say) A class rotdof2kfdof that held offsets and other info as class members?